src/bestcrypt_ve_fmt_plug.c

```diff
@@ -0,0 +1,456 @@
+
+//	assert(in && out && key && ivec);
+	while (len) {
+		unsigned char c;
+		serpent_decrypt(in, tmp.c, key);
+		for (n=0; n<16 && n<len; ++n) {
+			c = in[n];
+			out[n] = tmp.c[n] ^ ivec[n];
+			ivec[n] = c;
+		}
+		if (len<=16) {
+			for (; n<16; ++n)
+				ivec[n] = in[n];
+			break;
+		}
+		len -= 16;
+		in  += 16;
+		out += 16;
+	}
+}
+
+int Twofish_Decrypt_no_padding(Twofish_key *m_key, Byte *pInput, Byte *pOutBuffer,
+                               int nInputOctets, Byte *m_pInitVector)
+{
+	int i, numBlocks;
+	UInt32 iv[4];
+	union {
+		Byte block[16];
+		UInt32 p32[4];	// needed for 'requires aligned' machines
+	} x;
+	UInt32 *p;
+	Byte *block;
+
+	p = x.p32;
+	block = x.block;
+	if ((pInput == NULL) || (nInputOctets <= 0) || (pOutBuffer == NULL)) return 0;
+
+	if ((nInputOctets % 16) != 0) { return -1; }
+
+	numBlocks = nInputOctets / 16;
+
+	memcpy(iv, m_pInitVector, 16);
+
+	for (i = numBlocks; i > 0; i--)
+	{
+		Twofish_decrypt(m_key, (Twofish_Byte *)pInput, (Twofish_Byte *)block);
+		p[0] ^= iv[0];
+		p[1] ^= iv[1];
+		p[2] ^= iv[2];
+		p[3] ^= iv[3];
+		memcpy(iv, pInput, 16);
+		memcpy(pOutBuffer, block, 16);
+		pInput += 16;
+		pOutBuffer += 16;
+	}
+
+	return 16*numBlocks;
+}
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+
+#ifdef _OPENMP
+	max_threads = omp_get_max_threads();
+#else
+	max_threads = 1;
+#endif
+
+	local = mem_alloc(sizeof(*local) * max_threads);
+	int i;
+	for (i = 0; i < max_threads; i++)
+		yescrypt_init_local(&local[i]);
+
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	saved_len = mem_calloc(self->params.max_keys_per_crypt, sizeof(*saved_len));
+	cracked = mem_calloc(sizeof(*cracked), self->params.max_keys_per_crypt);
+	cracked_count = self->params.max_keys_per_crypt;
+
+	Twofish_initialise();
+}
+
+static void done(void)
+{
+	int i;
+	for (i = 0; i < max_threads; i++)
+		yescrypt_free_local(&local[i]);
+	MEM_FREE(local);
+
+	MEM_FREE(saved_key);
+	MEM_FREE(cracked);
+	MEM_FREE(saved_len);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *p = ciphertext, *ctcopy, *keeptr;
+	int extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LENGTH))
+		return 0;
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "$")) == NULL) // version
+		goto err;
+	if (!isdec(p))
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL) // encryption alogrithm
+		goto err;
+	if (!isdec(p))
+		goto err;
+	if (atoi(p) != 8 && atoi(p) != 9 && atoi(p) != 10 && atoi(p) != 11 && atoi(p) != 15)
+		goto err;
+	if (atoi(p) == 11) {
+		fprintf(stderr, "RC6 encryption not supported yet!\n");
+		goto err;
+	}
+	if ((p = strtokm(NULL, "$")) == NULL) // salt
+		goto err;
+	if (hexlenl(p, &extra) != 48 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL) // encrypted content
+		goto err;
+	if (hexlenl(p, &extra) != 0x60 * 2 || extra)
+		goto err;
```

Comment:
Should base the 48 and 0x60 here on the corresponding `sizeof` of struct fields.